### PR TITLE
Account operations source api

### PR DESCRIPTION
**Pull request by filipecabaco**
Adds multiple operations to an accounts source API 

* GET `/api/sources/:token` - Creates a new source to a given user
* POST `/api/sources` - Creates a new source to a given user
* PATCH `/api/sources/:token` - Updates a given source
* DELETE `/api/sources/:token` - Deletes a given source

### CODE REVIEW

File: lib/logflare_web/controllers/api/fallback_controller.ex [hunk 1]
```diff
@@ -0,0 +1,18 @@
+defmodule LogflareWeb.Api.FallbackController do
+  use Phoenix.Controller
+  alias Ecto.Changeset
+
+  def call(conn, {:error, %Changeset{} = changeset}) do
+    errors = Changeset.traverse_errors(changeset, fn _, _, {message, _} -> "#{message}" end)
```
Comment (Ziinc) [inline lib/logflare_web/controllers/api/fallback_controller.ex:6]: the `message` will already be a string, no need to type convert.

File: lib/logflare_web/controllers/api/source_controller.ex [hunk 2]
```diff
@@ -3,9 +3,38 @@ defmodule LogflareWeb.Api.SourceController do
...
+  action_fallback LogflareWeb.Api.FallbackController
+
   def index(%{assigns: %{user: user}} = conn, _) do
     user = Users.preload_sources(user)
     sources = Sources.preload_for_dashboard(user.sources)
     json(conn, sources)
   end
+
+  def create(%{assigns: %{user: user}} = conn, params) do
+    with params <- Map.put(params, "token", Ecto.UUID.generate()),
```
Comment (Ziinc) [inline lib/logflare_web/controllers/api/source_controller.ex:15]: we should shift down uuid generation to changeset level, actually. or even better, autogenerate in schema.

File: test/logflare_web/controllers/api/source_controller_test.exs [hunk 3]
```diff
@@ -42,4 +45,107 @@ defmodule LogflareWeb.Api.SourceControllerTest do
...
+    test "creates a new source for an authenticated user", %{
+      conn: conn,
+      users: [u1, _u2]
+    } do
+      name = TestUtils.random_string()
+
+      conn
+      |> login_user(u1)
+      |> post("/api/sources", %{name: name})
+      |> response(204)
```
Comment (Ziinc) [inline test/logflare_web/controllers/api/source_controller_test.exs:59]: shouldnt this be 201?

File: test/logflare_web/controllers/api/source_controller_test.exs [hunk 4]
```diff
@@ -42,4 +45,107 @@ defmodule LogflareWeb.Api.SourceControllerTest do
...
+
+      assert Repo.one(
```
Comment (Ziinc) [inline test/logflare_web/controllers/api/source_controller_test.exs:61]: should assert on response instead, controller layer shouldn't have repo knowledge.

File: test/logflare_web/controllers/api/source_controller_test.exs [hunk 5]
```diff
@@ -42,4 +45,107 @@ defmodule LogflareWeb.Api.SourceControllerTest do
...
+    } do
+      name = TestUtils.random_string()
+
+      conn
+      |> login_user(u1)
+      |> patch("/api/sources/#{s1.id}", %{name: name})
+      |> response(204)
+
+      assert name ==
+               Repo.one(
```
Comment (Ziinc) [inline test/logflare_web/controllers/api/source_controller_test.exs:85]: as above, can pattern match on response instead.

Comment (Ziinc) [reply test/logflare_web/controllers/api/source_controller_test.exs:85]: usual created resonse would be the return the created object

File: test/logflare_web/controllers/api/source_controller_test.exs [hunk 6]
```diff
@@ -42,4 +45,107 @@ defmodule LogflareWeb.Api.SourceControllerTest do
...
+      sources: [_s1, _s2, s3]
+    } do
+      conn
+      |> login_user(u1)
+      |> delete("/api/sources/#{s3.id}", %{name: TestUtils.random_string()})
+      |> response(404)
+    end
+  end
+
+  test "changeset errors handled gracefully", %{
```
Comment (Ziinc) [inline test/logflare_web/controllers/api/source_controller_test.exs:139]: we can merge this test together with update/2 and create/2 blocks, just adds 1 extra line of assertion for each test case.

```
# in setup
conn = login_user(user)


assert %{"error"=> %{"name"=> [ "can't" <>_]}} = conn 
        |> post("/api/sources" , @invalid) 
        |> json_response(422)


assert %{"error"=> %{"name"=> [ "can't" <>_]}} = conn 
        |> put("/api/sources/${s.id}" , @invalid) 
        |> json_response(422)

**Review by Ziinc — COMMENTED** [review]
another point to note is that throughout the app, `source.id` is used for routing and internal source identification, however, from an external user POV, they identify sources using the `source.token` uuid, as it is easier to remember the char strings start/endings.
This pattern is in line with the Logflare Endpoints api path params, where the `endpoint_query.token` is used in the path.
we should keep the api consistent and use uuids for external user params.

File: lib/logflare_web/controllers/api/fallback_controller.ex (hunk 1 above)
Comment (filipecabaco) [reply lib/logflare_web/controllers/api/fallback_controller.ex:6]: nice call! I had a different message being built 🤦

File: lib/logflare_web/controllers/api/source_controller.ex (hunk 2 above)
Comment (filipecabaco) [reply lib/logflare_web/controllers/api/source_controller.ex:15]: will change it 👍

File: test/logflare_web/controllers/api/source_controller_test.exs (hunk 3 above)
Comment (filipecabaco) [reply test/logflare_web/controllers/api/source_controller_test.exs:59]: oh yes! way better

File: test/logflare_web/controllers/api/source_controller_test.exs (hunk 4 above)
Comment (filipecabaco) [reply test/logflare_web/controllers/api/source_controller_test.exs:61]: the idea here was to just create the entity and reply an empty body. I can change it to have the entity in the response body and assert on it.

File: test/logflare_web/controllers/api/source_controller_test.exs (hunk 6 above)
Comment (filipecabaco) [reply test/logflare_web/controllers/api/source_controller_test.exs:139]: will change 👍

**Comment by filipecabaco** [comment]
Made requested changes. We're also using token instead of id to identify sources in the URL

**Review by chasers — APPROVED** [review]

**Pull request merged**